n28956: add notes and questions (#737)
bitcoin-core-review-club/website: _posts/2024-01-03-#28956.md

components: ["p2p", "consensus"]
host: stickies-v
status: upcoming
commit: 87c577e9742d7154826c755a7fe320f34fd54c81
---

## Notes

- Bitcoin relies on the concept of time for operations such as the difficulty adjustment and transaction- or script-level timelocks. Since time is relative and clocks are notoriously difficult to keep synchronized, it is impossible to define a source of truth that does not depend on synchronization and authorities such as NTP servers. In a decentralized system, we must accept that nodes can have a different but equally valid view of the current time.

- The Bitcoin network comes to consensus on time by requiring miners to commit to a timestamp in the [block header](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch09.asciidoc#block-header). To prevent miners from being able to use excessively deviating timestamps in their blocks, nodes [verify](https://github.com/bitcoin/bitcoin/blob/3a0f54dd2402d74a5ac4304b3ad09014cfb25edf/src/validation.cpp#L3803-L3810) that the timestamp is later than the `Median Time Past` (calculated on the previous 11 blocks), and earlier than 2 hours from the current time.

> - Prior to this PR, the current time was calculated based on the network-adjusted time. During the handshake, the peer node's current time is compared to the system's current time, and the time difference is [stored](https://github.com/bitcoin/bitcoin/blob/3a0f54dd2402d74a5ac4304b3ad09014cfb25edf/src/net_processing.cpp#L3556). The network-adjusted time is then calculated by adding the median of these offsets to the system's current time, unless it exceeds the `-maxtimeadjustment` value [in which case a warning is emitted](https://github.com/bitcoin/bitcoin/blob/3a0f54dd2402d74a5ac4304b3ad09014cfb25edf/src/timedata.cpp#L84-L100).

Reviewer: This could clarify what is peer-specific and how that becomes our view of the network.

```suggestion
- Prior to this PR, the current time was calculated based on the network-adjusted time. When connecting to a peer, the peer's current time is compared to the system's current time, and the time difference is [stored](https://github.com/bitcoin/bitcoin/blob/3a0f54dd2402d74a5ac4304b3ad09014cfb25edf/src/net_processing.cpp#L3556). The network-adjusted time is then calculated by adding the median of all peers' offsets to the system's current time, unless it exceeds the `-maxtimeadjustment` value [in which case a warning is emitted](https://github.com/bitcoin/bitcoin/blob/3a0f54dd2402d74a5ac4304b3ad09014cfb25edf/src/timedata.cpp#L84-L100).
```

Author: Thanks, I'll include this when I update the notes after the PR is rebased!

Author: Included in https://github.com/bitcoin-core-review-club/website/pull/738

- With this PR, the upper bound of the validity of a block header's timestamp is no longer calculated based on the network-adjusted time, but on the unadjusted system's current time. The network-adjusted time is still calculated and used to warn users of a potential clock misconfiguration.

- Note: the description in `timedata.cpp` [mentions that](https://github.com/bitcoin/bitcoin/blob/3a0f54dd2402d74a5ac4304b3ad09014cfb25edf/src/timedata.cpp#L66-L68) "the condition to update nTimeOffset includes checking whether the number of elements in vTimeOffsets is odd, which will never happen after there are 200 elements.". This comment has become outdated since [#6545](https://github.com/bitcoin/bitcoin/pull/6545), which made the maximum number of elements explicit in an earlier check.

## Questions

### Concept

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

2. Is it necessary for block headers to have a timestamp? If so, why?

> 3. What is the difference between Median Time Past (MTP) and network-adjusted time? Which of these are relevant to the PR?

Reviewer: ```suggestion
3. What is the difference between Median Time Past (MTP) and network-adjusted time? Which of these is relevant to the PR?
```

Author: It could be both, so I think "are" is the correct form?
